dkg/frost.go

```diff
@@ -0,0 +1,275 @@
+func newFrostParticipants(numValidators, numNodes, threshold, shareIdx uint32, dgkCtx string) (map[uint32]*frost.DkgParticipant, error) {
+	var otherIDs []uint32
+	for i := uint32(1); i <= numNodes; i++ {
+		if i == shareIdx {
+			continue
+		}
+		otherIDs = append(otherIDs, i)
+	}
+
+	resp := make(map[uint32]*frost.DkgParticipant)
+	for i := uint32(0); i < numValidators; i++ {
+		p, err := frost.NewDkgParticipant(
+			shareIdx,
+			threshold,
+			dgkCtx,
+			curves.BLS12381G1(),
+			otherIDs...)
+		if err != nil {
+			return nil, errors.Wrap(err, "new participant")
+		}
+
+		resp[i] = p
+	}
+
+	return resp, nil
+}
+
+// round1 executes round 1 for each validator and returns all round 1
+// broadcast and p2p messages for all validators.
+func round1(validators map[uint32]*frost.DkgParticipant) (map[msgKey]frost.Round1Bcast, map[msgKey]sharing.ShamirShare, error) {
+	var (
+		castResults = make(map[msgKey]frost.Round1Bcast)
+		p2pResults  = make(map[msgKey]sharing.ShamirShare)
+	)
+	for vIdx, v := range validators {
+		cast, p2p, err := v.Round1(nil)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "exec round 1")
+		}
+
+		castResults[msgKey{
+			ValIdx:   vIdx,
+			SourceID: v.Id,
+			TargetID: 0, // Broadcast
+		}] = *cast
+
+		for targetID, shamirShare := range p2p {
+			p2pResults[msgKey{
+				ValIdx:   vIdx,
+				SourceID: v.Id,
+				TargetID: targetID,
+			}] = *shamirShare
+		}
+	}
+
+	return castResults, p2pResults, nil
+}
+
+// round2 executes round 2 for each validator and returns all round 2
+// broadcast messages for all validators.
+func round2(
+	validators map[uint32]*frost.DkgParticipant,
+	castR1 map[msgKey]frost.Round1Bcast,
+	p2pR1 map[msgKey]sharing.ShamirShare,
+) (map[msgKey]frost.Round2Bcast, error) {
+	castResults := make(map[msgKey]frost.Round2Bcast)
+	for vIdx, v := range validators {
+		// Extract vIdx'th validator round 2 inputs.
+		var (
+			castMap  = make(map[uint32]*frost.Round1Bcast)
+			shareMap = make(map[uint32]*sharing.ShamirShare)
+		)
+		for key, cast := range castR1 {
+			if key.ValIdx != vIdx {
+				continue
+			}
+			if key.TargetID != v.Id {
+				continue
+			}
+			cast := cast // Copy loop variable
+			castMap[key.SourceID] = &cast
+		}
+		for key, share := range p2pR1 {
+			if key.ValIdx != vIdx {
+				continue
+			}
+			if key.TargetID != v.Id {
+				continue
+			}
+			share := share // Copy loop variable
+			shareMap[key.SourceID] = &share
+		}
+
+		castR2, err := v.Round2(castMap, shareMap)
+		if err != nil {
+			return nil, errors.Wrap(err, "exec round 2")
+		}
+
+		castResults[msgKey{
+			ValIdx:   vIdx,
+			SourceID: v.Id,
+			TargetID: 0, // Broadcast
+		}] = *castR2
+	}
+
+	return castResults, nil
+}
+
+// makeShares returns a slice of shares (one for each validator) from the DKG participants and round 2 results.
+func makeShares(validators map[uint32]*frost.DkgParticipant, r2Result map[msgKey]frost.Round2Bcast) ([]share, error) {
+	targetID := validators[0].Id
```

Comment:
can u pls explain this line? why the 0th validator's id?

Reply:
```suggestion
	// Get our ID from any validator (they all have our ID). 
	targetID := validators[0].Id
```